### DISCUSSION TITLE
1Password op:// URL scheme support

### DISCUSSION THREAD

**Discussion by drmikecrowe** [xonsh/xonsh#5580] [Q&A] [answered]
Does anybody know of work going on with [1Password ENV vars](https://developer.1password.com/docs/cli/secrets-environment-variables/) to pull passwords in the environment when they are formatted like this?

```
op://Private/OpenAI-API-Key/api key
```

I have the following workaround in my rc.d, but I wonder if that's the best way to do it:

```py
old_get = __xonsh__.env.get

def get_val(k, d=None):
    v = old_get(k, d)
    if type(v) == str and v.startswith("op://"):
        return $(op read @(v)).strip()
    return v

__xonsh__.env.get = get_val
```

IMO, we also need to help with [1Password completions](https://developer.1password.com/docs/cli/reference/commands/completion) as well

**Comment by anki-code** [comment]
The name of env variable that can have this data can be any or certain e.g. only `$1PASS_*`?

Basically you can register variable or pattern with detyping:
```xsh
__xonsh__.env.register(re.compile('\\w*PASS$'), detype=lambda v: v+"!")

$QQQ_PASS = 'qwe'

$QQQ_PASS
# 'qwe'

echo $QQQ_PASS
# qwe!

env | grep QQQ
# QQQ_PASS=qwe!
```

**Reply by drmikecrowe** [reply]
Anything -- Per their documentation, when you do an `op run -- ...` it replaces all the `op://1password-url-spec-to-a secret` with the actual secret.

I did a little more experimentation, and this is supported out of the box.  This works as expected:

```sh
@ op run -- chatblade what is pi
─────────────────────────────────────────────────────────────────────────────────────────────────── user ───────────────────────────────────────────────────────────────────────────────────────────────────
what is pi
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
──────────────────────────────────────────────────────────────────────────────────────────────── assistant ─────────────────────────────────────────────────────────────────────────────────────────────────
Pi (π) is a mathematical constant that represents the ratio of a circle's circumference to its diameter. It is an irrational number, meaning it cannot be expressed as a simple fraction or as a finite
decimal. The value of pi is approximately 3.14159, but it is an infinite and non-repeating decimal. It is one of the most important and widely used mathematical constants in various fields of science,
engineering, and mathematics.
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

```

And this fails;

```sh
@ chatblade what is pi
openai error: Error code: 401 - {'error': {'message': 'Incorrect API key provided: op://Pri***********************-key. You can find your API key at https://platform.openai.com/account/api-keys.', 'type':
'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}
```

I _thought_ my workaround was solving this issue, but it is only partially.  The first call (`chatgpt` is a xonsh xontrib, so it is loaded in the current context):

```sh
@ chatgpt what is pi
Replacing op://Private/OpenAI-API-Key/api-key from 1password

ChatGPT:

    Pi is a mathematical constant that represents the ratio of a circle's circumference to its diameter. It is an irrational number, approximately equal to 3.14159. Pi is commonly denoted by the Greek letter π.
```

However, if I run an external app, my workaround doesn't intercept that `__xonsh__.env.get` call:
```sh
@ chatblade what is pi
openai error: Error code: 401 - {'error': {'message': 'Incorrect API key provided: op://Pri***********************-key. You can find your API key at https://platform.openai.com/account/api-keys.', 'type':
'invalid_request_error', 'param': None, 'code': 'invalid_api_key'}}
```

So, I guess my question is this:  Is there a way to accomplish this in a xontrib, or will this need to be in xonsh itself?  I'd love if I could create a xontrib that handled _all_ `op://` environment vars, but I'm not sure how to do that.

**Reply by anki-code** [reply]
What about:
```xsh
import re
key_pattern = re.compile('\\w*key$')

__xonsh__.env.register(key_pattern, detype=lambda v: $(op read @(v)).strip() if v.startswith("op://") else v) 
echo $invalid_api_key
__xonsh__.env.deregister(key_pattern)

```
you can implement this as [python context manager](https://book.pythontips.com/en/latest/context_managers.html#implementing-a-context-manager-as-a-class) to have sugar:
```xsh
with open_my_keys():  # __enter__ is register
    echo $invalid_api_key
# __exit__ is deregister
```

But as I see in the 1pass documentation the name of the variable can by any.

**Reply by anki-code** [reply]
So it looks like a new feature - postprocessing the detyped value.

**Reply by anki-code** [reply]
@drmikecrowe I'm thinking about workaround for now and see two cases:

1. If you want to patch getting the value in any situation you need to patch [__getitem__](https://github.com/xonsh/xonsh/blob/01dc8858e6de4e7e569be8318288debe5aa1ac84/xonsh/environ.py#L2264) externally but [it's bad and hard](https://stackoverflow.com/questions/57413453/is-it-possible-to-override-getitem-at-instance-level-in-python).

2. If you want to patch only cases when env variables becomes string in subprocess commands you can patch detype. But this will not cover the cases when env variable used in python code e.g. `func($VAR)`. The start point for doing this here:
    
    ```xsh
    __xonsh__.env.old_get_detyper = __xonsh__.env.get_detyper
    
    def my_detyper(*a, **kw):
        """Convert variable from its type to a string representation in subproc mode."""
        real_detype = __xonsh__.env.old_get_detyper(*a, **kw)
        if real_detype:
            def new_detype(*adt, **kwdt):
                val = real_detype(*adt, **kwdt)
                if isinstance(val, str):
                    val = val + ' DECRYPT'
                return val
            return new_detype
    
    __xonsh__.env.get_detyper = my_detyper

    $ASD="qwe"
    $ASD
    # 'qwe'
    str($ASD)
    # 'qwe'
    __xonsh__.env.get('ASD')
    # 'qwe'
    
    echo $ASD
    # qwe DECRYPT
    env | grep ASD
    # ASD=qwe DECRYPT
    showcmd echo $ASD
    # ['echo', 'qwe DECRYPT']

    $UPDATE_OS_ENVIRON=True
    import os
    os.environ['ASD']
    # 'qwe DECRYPT'

    ```

But again this mostly about patching. Please create new feature request in xonsh issue tracker for adding pluggable way to do this.

**Comment by anki-code** [comment, marked as answer]
@drmikecrowe I figured out how to solve this without patching. It's more elegant and manageable. Take a look:

```xsh
# xontrib.onepassword

class OnePass:
    def __init__(self, url):
        self.url = url
    def __repr__(self):
        if __xonsh__.env.get("ONEPASS_ENABLED", False):
            return "DECRYPTED"
        else:
            return self.url
```
now:
```xsh
# xontrib load onepassword

$VAR = OnePass("op://key")
$VAR
# 'op://key'
echo $VAR
# op://key

$ONEPASS_ENABLED = True

$VAR
# 'DECRYPTED'
echo $VAR
# DECRYPTED

```

**Reply by anki-code** [reply]
@drmikecrowe please take a look.

**Reply by anki-code** [reply]
Hey @drmikecrowe! I became a user of 1password and looking for a xontrib. Do you have something to share?

**Reply by drmikecrowe** [reply]
Sorry @anki-code -- I haven't returned to this in a while.  Let me put one together and submit
UPD: https://github.com/drmikecrowe/xontrib-1password
